### PR TITLE
Adds max_seq_len to recipes and updates unit tests

### PR DESCRIPTION
**Pull request by thomasjpfan**
#### Context
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (Expose new feature in recipe yaml)

Addes task 1 of https://github.com/pytorch/torchtune/issues/1311

#### Changelog
This PR adds `max_seq_len` to the recipes and adds unit tests for loading the config

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](../tests/torchtune) for any new functionality
- [ ] update [docstrings](../docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: recipes/configs/eleuther_evaluation.yaml [hunk 1]
```diff
@@ -21,6 +21,7 @@ checkpointer:
 tokenizer:
   _component_: torchtune.models.llama2.llama2_tokenizer
   path: /tmp/Llama-2-7b-hf/tokenizer.model
+  max_seq_len: null
```
Comment (RdoubleA) [inline recipes/configs/eleuther_evaluation.yaml:24]: `max_seq_length` is already defined below and is used instead of the one with the tokenizer. To be consistent, let's keep the max_seq_len here in the tokenizer, set it to 4096, remove the other one, and update the eleuther_eval recipe to use the tokenizer's max_seq_len. You simply would need to update the line here: https://github.com/pytorch/torchtune/blob/f9f75bb563ecae371492a9d49da4a9f514c081b3/recipes/eleuther_eval.py#L246

to use the tokenizer max_seq_len: `self._cfg.tokenizer.max_seq_len`

**Review by RdoubleA — COMMENTED** [review]
Sorry for the delay, appreciate you doing this, this certainly improves the usability of all our configs. A few things:
- It seems like the mistral and phi3 configs were missed entirely? Any reason for that?
- For `mistral/7B_full_ppo_low_memory.yaml` in particular, we'd probably want to change this line: https://github.com/pytorch/torchtune/blob/f9f75bb563ecae371492a9d49da4a9f514c081b3/recipes/configs/mistral/7B_full_ppo_low_memory.yaml#L36 to be consistent with the classifier, where max_seq_len=32768 (cc @SalmanMohammadi)

**Comment by SalmanMohammadi** [comment]
>     * For `mistral/7B_full_ppo_low_memory.yaml` in particular, we'd probably want to change this line: https://github.com/pytorch/torchtune/blob/f9f75bb563ecae371492a9d49da4a9f514c081b3/recipes/configs/mistral/7B_full_ppo_low_memory.yaml#L36
>        to be consistent with the classifier, where max_seq_len=32768 (cc @SalmanMohammadi)

Text completion datasets don't defer to the tokenizer for max seq length truncation right? If this is the case it's fine to leave as null - unless I'm missing something re why it should be consistent with the classifier?
In any case this is more of a recipe/task hyperparam than it is about your models (max_seq_length will be in [64, 2048] or however big you can fit). If you're doing summarization tasks you might want it to be longer, if you're doing single-turn chat you might want it to be shorter, etc.

**Comment by RdoubleA** [comment]
Hi @thomasjpfan, let us know if you still plan to carry this through or if we should wrap it up. Appreciate your work so far!

**Comment by thomasjpfan** [comment]
> It seems like the mistral and phi3 configs were missed entirely? Any reason for that?

I had a quick script to automate the new keyword argument and missed mistral and phi3

From https://github.com/pytorch/torchtune/pull/1363#issuecomment-2307063732, I kept the tokenizer in `mistral/7B_full_ppo_low_memory.yaml` to `max_seq_len: null`.

File: tests/torchtune/config/test_instantiate.py [hunk 2]
```diff
@@ -74,3 +78,53 @@ def test_instantiate(self, config, module):
         del config.test.dim
         actual = instantiate(config.test, 3)
         assert self.get_dim(actual) == 3
+
+
+ASSETS = Path(__file__).parent.parent.parent / "assets"
+
+
+class TestInstantiateTokenizer:
```
Comment (RdoubleA) [inline tests/torchtune/config/test_instantiate.py:86]: Appreciate you adding this test, but actually don't think this is necessary. This is simply testing that an object can be instantiated and has the correct parameters, but this is already covered by the other tests in this directory

**Review by RdoubleA — COMMENTED** [review]
Looks great! Just one comment and it's ready to go

File: tests/torchtune/config/test_instantiate.py (hunk 2 above)
Comment (thomasjpfan) [reply tests/torchtune/config/test_instantiate.py]: I added the test to mostly check for the null case, which we couldn't do with `RMSNorm`.

In [`4eeb1fd` (#1363)](https://github.com/pytorch/torchtune/pull/1363/commits/4eeb1fd99d6e54335ce492b7023233b764a2e6a1), I simplified the test to only check the null case.

**Review by RdoubleA — APPROVED** [review]
